File: rust/kernel-receiver/src/display/tui.rs
use std::sync::atomic::{AtomicBool, Ordering};
use std::sync::{Arc, RwLock};
use std::time::Duration;

use crossterm::event::{self, Event, KeyCode, KeyEventKind};
use ratatui::layout::{Constraint, Layout};
use ratatui::style::{Style, Stylize};
use ratatui::widgets::{Block, Borders, Paragraph, Row, Table};

use crate::config::Config;
use crate::stats::Stats;

fn format_signature_prefix(sig: &[u8; 8]) -> String {
    let hex: String = sig.iter().map(|b| format!("{:02x}", b)).collect();
    format!("{}..{}", &hex[..4], &hex[12..])
}

fn format_duration_short(d: Duration) -> String {
    let secs = d.as_secs();
    if secs < 60 {
        format!("{}s", secs)
    } else if secs < 3600 {
        format!("{}m{}s", secs / 60, secs % 60)
    } else {
        format!("{}h{}m", secs / 3600, (secs % 3600) / 60)
    }
}

pub fn run(
    config: &Config,
    stats: Arc<RwLock<Stats>>,
    shutdown: Arc<AtomicBool>,
) -> anyhow::Result<()> {
    let tick_rate = Duration::from_millis(1000 / config.display.refresh_hz as u64);
    let mut terminal = ratatui::init();

    let result = run_loop(&mut terminal, &stats, &shutdown, tick_rate, config);

    ratatui::restore();
    result
}

fn run_loop(
    terminal: &mut ratatui::DefaultTerminal,
    stats: &Arc<RwLock<Stats>>,
    shutdown: &Arc<AtomicBool>,
    tick_rate: Duration,
    config: &Config,
) -> anyhow::Result<()> {
    while !shutdown.load(Ordering::Relaxed) {
        terminal.draw(|frame| {
            let chunks = Layout::vertical([
                Constraint::Length(3),
                Constraint::Fill(1),
                Constraint::Length(3),
            ])
            .split(frame.area());

            let mut s = stats.write().unwrap();

            // === Top bar ===
            let uptime = format_duration_short(s.start_time.elapsed());
            let hb_info = match s.last_heartbeat {
                Some(t) => format!(
                    "heartbeats: {} (last: {}ms ago)",
                    s.total_heartbeats,
                    t.elapsed().as_millis()
                ),
                None => format!("heartbeats: {} (none yet)", s.total_heartbeats),
            };
            let status_text = format!(
                " iface: {} | group: {} | uptime: {} | {}",
                config.network.interface, config.network.multicast_group, uptime, hb_info,
            );
            let status = Paragraph::new(status_text).block(
                Block::default()
                    .borders(Borders::ALL)
                    .title(" Edge Multicast Receiver "),
            );
            frame.render_widget(status, chunks[0]);

            // === Slot table ===
            let header = Row::new(vec![
                "Slot", "Signature", "Data", "Coding", "FEC Sets", "Age",
            ])
            .style(Style::new().bold());

            let rows: Vec<Row> = s
                .recent_slots()
                .iter()
                .map(|slot| {
                    let age = format_duration_short(slot.first_seen.elapsed());
                    Row::new(vec![
                        slot.slot.to_string(),
                        format_signature_prefix(&slot.signature_prefix),
                        slot.data_shred_count.to_string(),
                        slot.coding_shred_count.to_string(),
                        slot.fec_set_count.to_string(),
                        age,
                    ])
                })
                .collect();

            let table = Table::new(
                rows,
                [
                    Constraint::Length(12),
                    Constraint::Length(14),
                    Constraint::Length(8),
                    Constraint::Length(8),
                    Constraint::Length(10),
                    Constraint::Length(8),
                ],
            )
            .header(header)
            .block(
                Block::default()
                    .borders(Borders::ALL)
                    .title(" Recent Slots "),
            );

            frame.render_widget(table, chunks[1]);

            // === Bottom stats ===
            let rate = s.shreds_per_second();
            let total = s.total_data_shreds + s.total_coding_shreds;
            let ratio = if s.total_coding_shreds > 0 {
                format!(
                    "{:.1}",
                    s.total_data_shreds as f64 / s.total_coding_shreds as f64
                )
            } else {
                "n/a".into()
            };
            let stats_text = format!(
                " shreds/sec: {:.0} | total: {} (data: {}, coding: {}) | data/coding: {} | errors: {}",
                rate, total, s.total_data_shreds, s.total_coding_shreds, ratio, s.parse_errors,
            );
            let stats_bar = Paragraph::new(stats_text)
                .block(Block::default().borders(Borders::ALL).title(" Stats "));
            frame.render_widget(stats_bar, chunks[2]);
        })?;

        if event::poll(tick_rate)? {
            if let Event::Key(key) = event::read()? {
                if key.kind != KeyEventKind::Press {
                    continue;
                }
                match key.code {
                    KeyCode::Char('q') | KeyCode::Esc => {
                        shutdown.store(true, Ordering::Relaxed);
                        return Ok(());
                    }
                    _ => {}
                }
            }
        }
    }

    Ok(())
}

File: go/internal/display/tui.go
package display

import (
	"context"
	"fmt"
	"strings"
	"sync"
	"time"

	tea "github.com/charmbracelet/bubbletea"
	"github.com/charmbracelet/lipgloss"
	"github.com/malbeclabs/edge-multicast-ref/go/internal/stats"
)

// tickMsg triggers a periodic stats refresh in the TUI.
type tickMsg time.Time

// TUIModel is the bubbletea model for the stats dashboard.
type TUIModel struct {
	stats    *stats.Stats
	mu       *sync.RWMutex
	cancel   context.CancelFunc
	width    int
	height   int
	tickRate time.Duration

	// ExtraPanel is optional additional panel content (e.g. XDP stats).
	ExtraPanel string

	// Config values shown in the top bar.
	InterfaceName  string
	MulticastGroup string
}

// NewTUIModel creates a new TUI model.
func NewTUIModel(
	s *stats.Stats,
	mu *sync.RWMutex,
	cancel context.CancelFunc,
	refreshHz int,
	ifaceName, mcastGroup string,
) TUIModel {
	return TUIModel{
		stats:          s,
		mu:             mu,
		cancel:         cancel,
		tickRate:       time.Duration(1000/refreshHz) * time.Millisecond,
		InterfaceName:  ifaceName,
		MulticastGroup: mcastGroup,
	}
}

// Init returns the initial command (a periodic tick).
func (m TUIModel) Init() tea.Cmd {
	return tea.Tick(m.tickRate, func(t time.Time) tea.Msg { return tickMsg(t) })
}

// Update handles messages (key presses, window resize, tick).
func (m TUIModel) Update(msg tea.Msg) (tea.Model, tea.Cmd) {
	switch msg := msg.(type) {
	case tea.KeyMsg:
		switch msg.String() {
		case "q", "esc", "ctrl+c":
			m.cancel()
			return m, tea.Quit
		}
	case tea.WindowSizeMsg:
		m.width = msg.Width
		m.height = msg.Height
	case tickMsg:
		return m, tea.Tick(m.tickRate, func(t time.Time) tea.Msg { return tickMsg(t) })
	}
	return m, nil
}

// View renders the full TUI layout: top bar, slot table, bottom stats bar.
func (m TUIModel) View() string {
	if m.width == 0 {
		return "Initializing..."
	}

	m.mu.RLock()
	defer m.mu.RUnlock()

	innerWidth := m.width - 2 // account for border left+right

	borderStyle := lipgloss.NewStyle().
		Border(lipgloss.NormalBorder()).
		BorderForeground(lipgloss.Color("240")).
		Width(innerWidth)

	titleStyle := lipgloss.NewStyle().Bold(true)

	// ── Top bar ──────────────────────────────────────────────────────
	uptime := FormatDurationShort(time.Since(m.stats.StartTime))
	hbInfo := fmt.Sprintf("heartbeats: %d", m.stats.TotalHeartbeats)
	if m.stats.LastHeartbeat != nil {
		hbInfo += fmt.Sprintf(" (last: %dms ago)", time.Since(*m.stats.LastHeartbeat).Milliseconds())
	} else {
		hbInfo += " (none yet)"
	}
	topContent := fmt.Sprintf(" iface: %s | group: %s | uptime: %s | %s",
		m.InterfaceName, m.MulticastGroup, uptime, hbInfo)

	topBar := borderStyle.Render(
		titleStyle.Render(" Edge Multicast Receiver ") + "\n" + topContent,
	)

	// ── Bottom stats bar ─────────────────────────────────────────────
	rate := m.stats.ShredsPerSecond()
	total := m.stats.TotalDataShreds + m.stats.TotalCodingShreds
	ratio := "n/a"
	if m.stats.TotalCodingShreds > 0 {
		ratio = fmt.Sprintf("%.1f", float64(m.stats.TotalDataShreds)/float64(m.stats.TotalCodingShreds))
	}
	bottomContent := fmt.Sprintf(
		" shreds/sec: %.0f | total: %d (data: %d, coding: %d) | data/coding: %s | errors: %d",
		rate, total, m.stats.TotalDataShreds, m.stats.TotalCodingShreds, ratio, m.stats.ParseErrors,
	)
	bottomBar := borderStyle.Render(
		titleStyle.Render(" Stats ") + "\n" + bottomContent,
	)

	// ── Slot table ───────────────────────────────────────────────────
	header := fmt.Sprintf(" %-12s %-14s %8s %8s %10s %8s",
		"Slot", "Signature", "Data", "Coding", "FEC Sets", "Age")

	// Compute how many data rows we can fit.
	topH := lipgloss.Height(topBar)
	bottomH := lipgloss.Height(bottomBar)
	extraH := 0
	if m.ExtraPanel != "" {
		extraH = lipgloss.Height(m.ExtraPanel) + 1 // +1 for newline separator
	}
	// 4 = slot-table border (top+bottom) + header line + newline between panels
	usedHeight := topH + bottomH + extraH + 4
	maxRows := m.height - usedHeight
	if maxRows < 1 {
		maxRows = 1
	}

	slots := m.stats.RecentSlots()
	var tableRows strings.Builder
	tableRows.WriteString(header)
	tableRows.WriteByte('\n')
	for i, ss := range slots {
		if i >= maxRows {
			break
		}
		age := FormatDurationShort(time.Since(ss.FirstSeen))
		fmt.Fprintf(&tableRows, " %-12d %-14s %8d %8d %10d %8s\n",
			ss.Slot, FormatSignaturePrefix(ss.SignaturePrefix),
			ss.DataShredCount, ss.CodingShredCount,
			ss.FECSetCount, age)
	}

	slotTable := borderStyle.Render(
		titleStyle.Render(" Recent Slots ") + "\n" + tableRows.String(),
	)

	// ── Compose layout ───────────────────────────────────────────────
	var out strings.Builder
	out.WriteString(topBar)
	out.WriteByte('\n')
	if m.ExtraPanel != "" {
		out.WriteString(m.ExtraPanel)
		out.WriteByte('\n')
	}
	out.WriteString(slotTable)
	out.WriteByte('\n')
	out.WriteString(bottomBar)
	return out.String()
}

// RunTUI starts the bubbletea TUI program. It blocks until the user quits or
// the context is cancelled.
func RunTUI(ctx context.Context, s *stats.Stats, mu *sync.RWMutex, refreshHz int, ifaceName, mcastGroup string) error {
	ctx, cancel := context.WithCancel(ctx)
	defer cancel()

	model := NewTUIModel(s, mu, cancel, refreshHz, ifaceName, mcastGroup)
	p := tea.NewProgram(model, tea.WithAltScreen())

	done := make(chan error, 1)
	go func() {
		_, err := p.Run()
		done <- err
		cancel()
	}()

	select {
	case <-ctx.Done():
		p.Quit()
		return nil
	case err := <-done:
		return err
	}
}
